### ISSUE TITLE
Flash: does not work in scenarios with rendered response

### ISSUE THREAD

**Issue by christophs78** [eclipse-ee4j/mojarra#4977]
When we try to pass Java-objects via Flash to a PrimeFaces Dialog Framework (which is technically a IFRAME; http://primefaces.org/showcase/ui/df/data.xhtml) this does not work due to 
`faces.externalcontext.flash.response.already.committed=JSF1095: The response was already committed by the time we tried to set the outgoing cookie for the flash.  Any values stored to the flash will not be available on the next request.`

Debugging (with Mojarra 2.3.16) show´s following:

`FaceletViewHandlingStrategy#renderView`  calls via some intermediate steps `HtmlResponseWriter#endDocument`. And `HtmlResponseWriter#endDocument` calls `flush`.
![Mojarra_Flash_01_Flush](https://user-images.githubusercontent.com/10461942/137467550-bc824a58-e6f5-4392-a5e3-d96d3781e44e.png)

After this `FaceletViewHandlingStrategy#renderView`  calls `ctx.getExternalContext().getFlash().doPostPhaseActions(ctx);` which want´s to set the Flash-cookie. This can´t work because of the flush before.
![Mojarra_Flash_02_Flush](https://user-images.githubusercontent.com/10461942/137468546-8f7d4ef7-850b-43c7-966d-637d5991b148.png)

Ironically this works when you use https://showcase.omnifaces.org/filters/GzipResponseFilter because this filter kind of ignores flush. (We used GzipResponseFilter up to now. After switching to GZIP provided by the container our existing usecases broke.)

(MyFaces does not have this issue.)

And i´ll try to provide a PR. So you may assign this issue to me.

**Comment by BalusC** [comment]
Wondering about this comment in `FaceletViewHandlingStrategy#renderView()`
```
//  Don't call startDoc and endDoc on a partial response
```
I digged in the commit history and found this commit: https://github.com/javaee/mojarra/commit/6de4203e2ee9a0ced8198b53ab6a36e8818cb152

The currently reported bug was actually introduced over there.

The `startDocument()` (and `endDocument()`) was moved to `PartialViewContextImpl#processPartial()` because it's also setting response headers *after* setting the partial response writer like so:
```
exContext.setResponseContentType(RIConstants.TEXT_XML_CONTENT_TYPE);
exContext.addResponseHeader("Cache-Control", "no-cache");
```
And that didn't work when `startDocument()` is invoked before that. It was actually not necessary to move `endDocument()` to there as well, I imagine that it was done for clarity. But actually the `Flash#doPostPhaseActions()` should during that commit *also* have been moved to `PartialViewContextImpl`. This was not done and is causing the currently reported issue.

I'll create an improved PR.
